scripts/update_appup.escript

```diff
@@ -1,75 +1,313 @@
+    New = New0 -- AlreadyHandled,
+    Changed = Changed0 -- AlreadyHandled,
+    Deleted = Deleted0 -- AlreadyHandled,
+    [{load_module, M, brutal_purge, soft_purge, []} || M <- Changed ++ New] ++
+        OldActions ++
+        [{delete_module, M} || M <- Deleted].
+
+ensure_pred_versions(PredVersion, Versions) ->
+    {Maj, Min, Patch} = parse_semver(PredVersion),
+    PredVersions = [semver(Maj, Min, P) || P <- lists:seq(0, Patch)],
+    lists:foldl(fun ensure_version/2, Versions, PredVersions).
 
-add_modules(NewModules, OldActions) ->
-    OldModules = lists:map(fun(It) -> element(2, It) end, OldActions),
-    Modules = NewModules -- OldModules,
-    OldActions ++ [{load_module, M, brutal_purge, soft_purge, []} || M <- Modules].
+ensure_version(Version, Versions) ->
+    case lists:keyfind(Version, 1, Versions) of
+        false ->
+            [{Version, []}|Versions];
+        _ ->
+            Versions
+    end.
 
 read_appup(File) ->
-    {ok, Bin0} = file:read_file(File),
-    %% Hack:
-    Bin1 = re:replace(Bin0, "VSN", "\"VSN\""),
-    TmpFile = filename:join("/tmp", filename:basename(File)),
-    ok = file:write_file(TmpFile, Bin1),
-    {ok, [Terms]} = file:consult(TmpFile),
-    Terms.
+    case file:script(File, [{'VSN', "VSN"}]) of
+        {ok, Terms} ->
+            Terms;
+        Error ->
+            fail("Failed to parse appup file ~s: ~p", [File, Error])
+    end.
+
+diff_releases(Curr, Old) ->
+    Fun = fun(App, Modules, Acc) ->
+                  OldModules = maps:get(App, Old, #{}),
+                  Acc#{App => diff_app_modules(Modules, OldModules)}
+          end,
+    maps:fold(Fun, #{}, Curr).
+
+diff_app_modules(Modules, OldModules) ->
+    {New, Changed} =
+        maps:fold( fun(Mod, MD5, {New, Changed}) ->
+                           case OldModules of
+                               #{Mod := OldMD5} when MD5 =:= OldMD5 ->
+                                   {New, Changed};
+                               #{Mod := _} ->
+                                   {New, [Mod|Changed]};
+                               _ -> {[Mod|New], Changed}
+                           end
+                   end
+                 , {[], []}
+                 , Modules
+                 ),
+    Deleted = maps:keys(maps:without(maps:keys(Modules), OldModules)),
+    {New, Changed, Deleted}.
+
+find_beams(Dir) ->
+    [filename:join(Dir, I) || I <- filelib:wildcard("**/ebin/*.beam", Dir)].
+
+prepare(Baseline, #{clone_url := Repo, make_command := MakeCommand, beams_dir := BeamDir}) ->
+    log("~n===================================~n"
+        "Baseline: ~s"
+        "~n===================================~n", [Baseline]),
+    log("Building the current version...~n"),
+    bash(MakeCommand),
+    log("Downloading and building the previous release...~n"),
+    {ok, PredRootDir} = build_pred_release(Baseline, Repo, MakeCommand),
+    {BeamDir, filename:join(PredRootDir, BeamDir)}.
+
+build_pred_release(Baseline, Repo, MakeCommand) ->
```

Comment:
just question, have you considered to use released zip package? 
such like 
https://github.com/emqx/emqx/releases/download/v4.3.8/emqx-centos7-4.3.8-amd64.zip

Reply:
I will add the support in the next version